lib/encodings/base16.fz

```diff
@@ -0,0 +1,104 @@
+# This file is part of the Fuzion language implementation.
+#
+# The Fuzion language implementation is free software: you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as published
+# by the Free Software Foundation, version 3 of the License.
+#
+# The Fuzion language implementation is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+# License for more details.
+#
+# You should have received a copy of the GNU General Public License along with The
+# Fuzion language implementation.  If not, see <https://www.gnu.org/licenses/>.
+
+
+# -----------------------------------------------------------------------
+#
+#  Tokiwa Software GmbH, Germany
+#
+#  Source code of Fuzion standard library feature base16
+#
+# -----------------------------------------------------------------------
+
+# base16 encoding and decoding as defined in RFC 4648
+# https://datatracker.ietf.org/doc/html/rfc4648#section-8
+#
+public base16 is
+
+  alphabet array u8:= [48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 65, 66, 67, 68, 69, 70] #0123456789ABCDEF
+  encoding_name => "base16"
+
+  # decode a valid base16 character to 4 bits
+  four_bits(n u8) =>
+    if 48 <= n <= 57        # case 0-9
+      n.as_u8 - 48
+    else #if 65 <= n <= 70  # case A-F
+      n.as_u8 - 55
+
+  # checks if a character is valid in the encoding
+  is_valid(c u8) =>
+    (48 <= c <= 57) || (65 <= c <= 70)
+
+  # Encodes a given byte sequence in base16
+  # returns a sequence of ascii values
+  public encode(data array u8) =>
+    data.flat_map u8 (b->[alphabet[(b>>(u8 4)).as_i32], alphabet[(b&(u8 15)).as_i32]])
```

Comment:
b>>(u8 4) => b>>4

Reply:
b&(u8 15) => b&15

Reply:
Hm here it does actually work, but I've had the case where b was of type u64 and there it didn't work.

Reply:
Never mind there I've used a field instead of a constant